File: config/pixel/lua/config/pixel/ts/libs/ui.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayIsArray = ____lualib.__TS__ArrayIsArray
local ____exports = {}
local ____core = require("config.pixel.ts.libs.core")
local applyZic = ____core.applyZic
function ____exports.getPosition(self, pos)
    return __TS__ArrayIsArray(pos) and table.concat(pos, " ") or pos
end
--- Create a component
-- 
-- @param name string - The name of the component to create.
-- @param position string | string[] - The position of the component.
-- @param values table - An array of Zic values to apply to the component.
-- @returns table
function ____exports.getComponent(self, name, position, values)
    if values == nil then
        values = {}
    end
    return {
        {COMPONENT = (name .. " ") .. ____exports.getPosition(nil, position)},
        unpack(values)
    }
end
--- Create a component
-- 
-- @param name string - The name of the component to create.
-- @param position string | string[] - The position of the component.
-- @param values table - An array of Zic values to apply to the component.
function ____exports.component(self, name, position, values)
    if values == nil then
        values = {}
    end
    applyZic(____exports.getComponent(nil, name, position, values))
end
--- Add a zone encoder
-- 
-- @param position string | string[] - The position of the component.
function ____exports.addZoneEncoder(self, position)
    applyZic({{ADD_ZONE_ENCODER = ____exports.getPosition(nil, position)}})
end
--- Create a view
-- 
-- @param name string - The name of the view to create.
-- @param values table - An array of Zic values to apply to the view.
-- @returns table
function ____exports.getView(self, name, values)
    if values == nil then
        values = {}
    end
    return {
        {VIEW = name},
        unpack(values)
    }
end
--- Create a view and apply Zic values to it.
-- 
-- @param name string - The name of the view to create.
-- @param values table - An array of Zic values to apply to the view.
function ____exports.view(self, name, values)
    if values == nil then
        values = {}
    end
    applyZic(____exports.getView(nil, name, values))
end
--- Set the window position
-- 
-- @param x number The x position
-- @param y number The y position
function ____exports.setWindowPosition(self, x, y)
    applyZic({{WINDOW_POSITION = (tostring(x) .. " ") .. tostring(y)}})
end
--- Set the screen size
-- 
-- @param width number The width of the screen
-- @param height number The height of the screen
function ____exports.setScreenSize(self, width, height)
    applyZic({{SCREEN = (tostring(width) .. " ") .. tostring(height)}})
end
return ____exports

File: src/config/libs/ui.ts
import { applyZic, jsonStringify, ZicValue } from './core';
import { ComponentProps } from './nativeComponents/component';

export type Bounds = string | string[] | number[];

// TODO to be deprecated
export function getPosition(pos: Bounds) {
    return Array.isArray(pos) ? pos.join(' ') : pos;
}

export function getBounds(pos: Bounds) {
    return Array.isArray(pos) ? pos : pos.split(' ');
}

export function getJsonComponent<P>(componentName: string) {
    return (props: ComponentProps<P>) => {
        const COMPONENT = jsonStringify({
            ...props,
            componentName,
            bounds: getBounds(props.bounds),
        });
        return [{ COMPONENT }];
    };
}

/**
 * Add a zone encoder
 * @param position string | string[] - The position of the component.
 */
export function addZoneEncoder(bounds: Bounds) {
    applyZic([{ ADD_ZONE_ENCODER: getPosition(bounds) }]);
}

/**
 * Create a view
 *
 * @param name string - The name of the view to create.
 * @param values table - An array of Zic values to apply to the view.
 * @returns table
 */
export function getView(name: string, values: ZicValue[] = []) {
    return [{ VIEW: name }, values];
}

/**
 * Create a view and apply Zic values to it.
 *
 * @param name string - The name of the view to create.
 * @param values table - An array of Zic values to apply to the view.
 */
export function view(name: string, values: ZicValue[] = []) {
    applyZic(getView(name, values));
}

/**
 * Set the window position
 *
 * @param x number The x position
 * @param y number The y position
 */
export function setWindowPosition(x: number, y: number) {
    applyZic([{ WINDOW_POSITION: `${x} ${y}` }]);
}

/**
 * Set the screen size
 * @param width number The width of the screen
 * @param height number The height of the screen
 */
export function setScreenSize(width: number, height: number) {
    applyZic([{ SCREEN: `${width} ${height}` }]);
}


/**
 * Returns a string representing a color in hexadecimal notation.
 *
 * @param r number The red component of the color, between 0 and 255.
 * @param g number The green component of the color, between 0 and 255.
 * @param b number The blue component of the color, between 0 and 255.
 *
 * @returns string A string in the format `#RRGGBB`, where `RR`, `GG`, and `BB` are the red, green, and blue components of the color, respectively, in hexadecimal notation.
 */
export function rgb(r: number, g: number, b: number) {
    return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
}

/**
 * Returns a string representing a color with an alpha channel in hexadecimal notation.
 *
 * @param r number The red component of the color, between 0 and 255.
 * @param g number The green component of the color, between 0 and 255.
 * @param b number The blue component of the color, between 0 and 255.
 * @param a number The alpha component of the color, between 0 and 1.
 *
 * @returns string A string in the format `#RRGGBBAA`, where `RR`, `GG`, `BB`, and `AA` are the red, green, blue, and alpha components of the color, respectively, in hexadecimal notation.
 */
export function rgba(r: number, g: number, b: number, a: number) {
    return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}${(Math.round(a * 255)).toString(16).padStart(2, '0')}`;
}